server/historian/packages/historian-base/src/routes/utils.ts

```diff
@@ -47,9 +62,147 @@ export class createGitServiceArgs {
+async function checkAndCacheIsEphemeral({
+	documentId,
+	tenantId,
+	documentManager,
+	ephemeralDocumentTTLSec,
+	isEphemeralContainerOverride,
+	cache,
+}: {
+	documentId: string;
+	tenantId: string;
+	documentManager: IDocumentManager;
+	ephemeralDocumentTTLSec: number;
+	isEphemeralContainerOverride?: boolean;
+	cache?: ICache;
+}): Promise<boolean> {
+	const isEphemeralKey = getEphemeralContainerCacheKey(documentId);
+	if (typeof isEphemeralContainerOverride === "boolean") {
+		// If an isEphemeralContainerOverride flag was passed in, cache it and return it.
+		// This typically happens on the first request to the document.
+		await updateIsEphemeralCache(documentId, tenantId, isEphemeralContainerOverride, cache);
+		return isEphemeralContainerOverride;
+	}
+	// When no override is provided, check the cache first.
+	const cachedIsEphemeral: boolean | undefined = await runWithRetry<boolean | undefined>(
+		async () => cache?.get(isEphemeralKey) /* api */,
+		"utils.createGitService.get" /* callName */,
+		3 /* maxRetries */,
+		1000 /* retryAfterMs */,
+		getLumberBaseProperties(documentId, tenantId) /* telemetryProperties */,
+	).catch((error) => {
+		Lumberjack.error(
+			`Error getting ${isEphemeralKey} from redis`,
+			getLumberBaseProperties(documentId, tenantId),
+			error,
+		);
+		return undefined;
+	});
+	if (typeof cachedIsEphemeral === "boolean") {
+		return cachedIsEphemeral;
+	}
+
+	// Finally, if isEphemeral was not in the cache, fetch the value from database.
+	const staticProps: IDocumentStaticProperties | undefined =
+		await runWithRetry<IDocumentStaticProperties>(
+			async () => documentManager.readStaticProperties(tenantId, documentId) /* api */,
+			"utils.createGitService.readStaticProperties" /* callName */,
+			3 /* maxRetries */,
+			1000 /* retryAfterMs */,
+			getLumberBaseProperties(documentId, tenantId) /* telemetryProperties */,
+		).catch((error) => {
+			Lumberjack.error(
+				`Failed to retrieve static data from document when checking isEphemeral flag.`,
+				getLumberBaseProperties(documentId, tenantId),
+				error,
+			);
+			return undefined;
+		});
+	if (staticProps?.isEphemeralContainer === undefined) {
+		Lumberjack.warning(
+			`No isEphemeralContainer flag found in static data for document.`,
+			getLumberBaseProperties(documentId, tenantId),
+		);
+		return false;
+	}
+	const staticPropsIsEphemeral = staticProps.isEphemeralContainer;
+	if (staticPropsIsEphemeral === true) {
+		// Explicitly check for ephemeral document expiration.
+		const currentTime = Date.now();
+		const documentExpirationTime = staticProps.createTime + ephemeralDocumentTTLSec * 1000;
+		if (currentTime > documentExpirationTime) {
+			// If the document is ephemeral and older than the max ephemeral document TTL, throw an error indicating that it can't be accessed.
+			const documentExpiredByMs = currentTime - documentExpirationTime;
+			const error = new NetworkError(
+				404,
+				`Ephemeral Container Expired: ${Math.floor(
```

Comment:
Can we move this to properties for more accurate error counting? Something like `timeSinceEphemeralContainerExpired`?

Reply:
Technically, this is the Network Error message that is returned to clients. It will also be logged, but I think it will be logged as the `exception` field, not the main message. Like `message: Request failed; exception: NetworkError...`